streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -228,8 +221,188 @@ public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeB
+                           We need to:
+                             1. Move the repartition to come immediately after the map call
+                             2. Remove the repartition between peek and groupByKey
+                             3. Re-attach graph children on either side to keep sub-topology in place
+                        */
+
+                    StreamsGraphNode keyChangingNodeChild = findParentNodeMatching(repartitionNodeToBeReplaced, gn -> gn.parentNode().equals(keyChangingNode));
+
+                    LOG.debug("Found the node downstream of key-changer {}", keyChangingNodeChild);
+
+                    // need to add children of key-changing node as children of optimized repartition
+                    // in order to process records from re-partitioning
+                    optimizedSingleRepartition.addChildNode(keyChangingNodeChild);
+
+                    LOG.debug("Removing {} from {}  children {}", keyChangingNodeChild, keyChangingNode, keyChangingNode.children());
+                    // now remove children from key-changing node
+                    keyChangingNode.removeChild(keyChangingNodeChild);
+
+                    // now need to get children of repartition node so we can remove repartition node
+                    Set<StreamsGraphNode> repartitionNodeToBeReplacedChildren = repartitionNodeToBeReplaced.children();
+
+                    StreamsGraphNode parentOfRepartitionNodeToBeReplaced = findParentNodeMatching(repartitionNodeToBeReplaced, n -> n.equals(repartitionNodeToBeReplaced.parentNode()));
+
+                    LOG.debug("Returned {} parentOfRepartitionNodeToBeReplaced as parent of {}", parentOfRepartitionNodeToBeReplaced, repartitionNodeToBeReplaced);
+
+                    // remove repartition node
+                    parentOfRepartitionNodeToBeReplaced.removeChild(repartitionNodeToBeReplaced);
+
+                    repartitionNodeToBeReplaced.clearChildren();
+
+                    // re-attach children of removed repartition
+                    for (StreamsGraphNode repartitionNodeToBeReplacedChild : repartitionNodeToBeReplacedChildren) {
+                        parentOfRepartitionNodeToBeReplaced.addChildNode(repartitionNodeToBeReplacedChild);
+                    }
+                }
+
+                LOG.debug("UPDATED node {} children {}", optimizedSingleRepartition, optimizedSingleRepartition.children());
+            }
+
+            keyChangingNode.addChildNode(optimizedSingleRepartition);
+            LOG.debug("Key Changing node {} children {}", keyChangingNode, keyChangingNode.children());
+        }
+
+        LOG.debug("Fully updated repartition {} children {}", optimizedSingleRepartition,
+                  optimizedSingleRepartition != null ? optimizedSingleRepartition.children() : null);
+    }
+
+    private StreamsGraphNode createRepartitionNode(String name, Serde keySerde, Serde valueSerde) {
```

Comment:
KGroupedTableImpl also have a duplicated `createRepartitionNode`.

Reply:
Ack will try to collapse the two in 4th PR, but for now, the `createReparitionNode` is removed from `InternalStreamsBuilder` until 4th PR